## ![s](https://s19.postimg.org/x69rfot6b/Pampa_ds160.png)

In [1]:
import pygsheets

In [2]:
import pandas as pd

In [3]:
gc = pygsheets.authorize(outh_file='GoogleSheets_client_secret_OAuth.json')


### Our sheet in Google Drive


### ![s](http://s19.postimg.org/myr83a6yb/Pampa_ds160.png)

In [4]:
%%time
# Find a workbook by name and open the first sheet
# Make sure you use the right name here.
sheet=gc.open('TEST_Emp_data').sheet1

Wall time: 676 ms


In [5]:
# Extract and print all of the values
TEST_Emp_data = sheet.get_all_records()
type(TEST_Emp_data)

list

In [6]:
TEST_Emp_data

[{'ID': 1,
  'Location': '2°32′S 140°43′E Jayapura, Indonesia',
  'Name': 'Spring Magoon'},
 {'ID': 2,
  'Location': '22°34′S 17°05′E Windhoek, Namibia',
  'Name': 'Gema Coberly'},
 {'ID': 3,
  'Location': '44°45′N 19°42′E Šabac, Serbia',
  'Name': 'Paige Wallace'},
 {'ID': 4, 'Location': '15°36′S 56°06′W Cuiabá, Brazil', 'Name': 'Sun Mees'},
 {'ID': 5,
  'Location': '36°23′N 139°04′E Maebashi, Japan',
  'Name': 'Germaine Manuelito'},
 {'ID': 6, 'Location': '13°32′S 71°58′W Cusco, Peru', 'Name': 'Iris Cowen'},
 {'ID': 7,
  'Location': '12°07′N 68°56′W Willemstad, Netherlands',
  'Name': 'Shantae Sulton'},
 {'ID': 8,
  'Location': '38°33′N 121°28′W Sacramento, United States',
  'Name': 'Tanika Carrera'},
 {'ID': 9,
  'Location': '31°56′N 35°56′E Amman, Jordan',
  'Name': 'Lisabeth Milburn'},
 {'ID': 10,
  'Location': '54°05′N 2°80′W Lancaster, United Kingdom',
  'Name': 'Carletta Slayton'}]

### Saving the Google Sheet into Pandas dataframe

In [28]:
df = pd.DataFrame.from_records(TEST_Emp_data)

#### changing the order of DataFrame columns

In [29]:
df=df[['ID','Name','Location']]

In [30]:
df

ID                Name                                    Location
0   1       Spring Magoon         2°32′S 140°43′E Jayapura, Indonesia
1   2        Gema Coberly           22°34′S 17°05′E Windhoek, Namibia
2   3       Paige Wallace               44°45′N 19°42′E Šabac, Serbia
3   4            Sun Mees              15°36′S 56°06′W Cuiabá, Brazil
4   5  Germaine Manuelito            36°23′N 139°04′E Maebashi, Japan
5   6          Iris Cowen                 13°32′S 71°58′W Cusco, Peru
6   7      Shantae Sulton     12°07′N 68°56′W Willemstad, Netherlands
7   8      Tanika Carrera  38°33′N 121°28′W Sacramento, United States
8   9    Lisabeth Milburn               31°56′N 35°56′E Amman, Jordan
9  10    Carletta Slayton    54°05′N 2°80′W Lancaster, United Kingdom

### Updated logic, Thank you to Kevin Markham <kevin@dataschool.io>

In [31]:
df['Address'] = df.Location.str.split(' ').apply(lambda x: ' '.join(x[2:]))

In [32]:
df

ID                Name                                    Location  \
0   1       Spring Magoon         2°32′S 140°43′E Jayapura, Indonesia   
1   2        Gema Coberly           22°34′S 17°05′E Windhoek, Namibia   
2   3       Paige Wallace               44°45′N 19°42′E Šabac, Serbia   
3   4            Sun Mees              15°36′S 56°06′W Cuiabá, Brazil   
4   5  Germaine Manuelito            36°23′N 139°04′E Maebashi, Japan   
5   6          Iris Cowen                 13°32′S 71°58′W Cusco, Peru   
6   7      Shantae Sulton     12°07′N 68°56′W Willemstad, Netherlands   
7   8      Tanika Carrera  38°33′N 121°28′W Sacramento, United States   
8   9    Lisabeth Milburn               31°56′N 35°56′E Amman, Jordan   
9  10    Carletta Slayton    54°05′N 2°80′W Lancaster, United Kingdom   

                     Address  
0        Jayapura, Indonesia  
1          Windhoek, Namibia  
2              Šabac, Serbia  
3             Cuiabá, Brazil  
4            Maebashi, Japan  
5                Cusco, Peru  
6    Willemstad, Netherlands  
7  Sacramento, United States  
8              Amman, Jordan  
9  Lancaster, United Kingdom

## we are getting location in the form of (2°32′S 140°43′E Jayapura, Indonesia)we'll make it more tidy

In [33]:
df['Country']=df.Location.str.split(',').apply(lambda x:x[1])

In [34]:
df['Latitude']=df.Location.str.split(' ').apply(lambda x:x[0])

In [35]:
df['Longitude']=df.Location.str.split(' ').apply(lambda x:x[1])

In [36]:
df['State']=df.Location.str.split(', ').apply(lambda x:x[0])

In [37]:
df['State']=df['State'].apply(lambda x:x.split(' ')[2])

In [15]:
"""def clean(x):
    x.pop(0)
    x.pop(0)
    return(x)"""

'def clean(x):\n    x.pop(0)\n    x.pop(0)\n    return(x)'

In [38]:
df=df[['ID','Name','State','Country','Latitude','Longitude','Address']]

In [39]:
df

ID                Name       State          Country Latitude Longitude  \
0   1       Spring Magoon    Jayapura        Indonesia   2°32′S  140°43′E   
1   2        Gema Coberly    Windhoek          Namibia  22°34′S   17°05′E   
2   3       Paige Wallace       Šabac           Serbia  44°45′N   19°42′E   
3   4            Sun Mees      Cuiabá           Brazil  15°36′S   56°06′W   
4   5  Germaine Manuelito    Maebashi            Japan  36°23′N  139°04′E   
5   6          Iris Cowen       Cusco             Peru  13°32′S   71°58′W   
6   7      Shantae Sulton  Willemstad      Netherlands  12°07′N   68°56′W   
7   8      Tanika Carrera  Sacramento    United States  38°33′N  121°28′W   
8   9    Lisabeth Milburn       Amman           Jordan  31°56′N   35°56′E   
9  10    Carletta Slayton   Lancaster   United Kingdom  54°05′N    2°80′W   

                     Address  
0        Jayapura, Indonesia  
1          Windhoek, Namibia  
2              Šabac, Serbia  
3             Cuiabá, Brazil  
4            Maebashi, Japan  
5                Cusco, Peru  
6    Willemstad, Netherlands  
7  Sacramento, United States  
8              Amman, Jordan  
9  Lancaster, United Kingdom

In [40]:
df.set_index('ID',inplace=True)

In [41]:
df

Name       State          Country Latitude Longitude  \
ID                                                                       
1        Spring Magoon    Jayapura        Indonesia   2°32′S  140°43′E   
2         Gema Coberly    Windhoek          Namibia  22°34′S   17°05′E   
3        Paige Wallace       Šabac           Serbia  44°45′N   19°42′E   
4             Sun Mees      Cuiabá           Brazil  15°36′S   56°06′W   
5   Germaine Manuelito    Maebashi            Japan  36°23′N  139°04′E   
6           Iris Cowen       Cusco             Peru  13°32′S   71°58′W   
7       Shantae Sulton  Willemstad      Netherlands  12°07′N   68°56′W   
8       Tanika Carrera  Sacramento    United States  38°33′N  121°28′W   
9     Lisabeth Milburn       Amman           Jordan  31°56′N   35°56′E   
10    Carletta Slayton   Lancaster   United Kingdom  54°05′N    2°80′W   

                      Address  
ID                             
1         Jayapura, Indonesia  
2           Windhoek, Namibia  
3               Šabac, Serbia  
4              Cuiabá, Brazil  
5             Maebashi, Japan  
6                 Cusco, Peru  
7     Willemstad, Netherlands  
8   Sacramento, United States  
9               Amman, Jordan  
10  Lancaster, United Kingdom

In [25]:
sh=gc.open('TEST_Emp_data')
wks = sh.sheet1

In [29]:
%%time
# create a new sheet with 50 rows and 60 colums
wks=sh.add_worksheet("Pandas Sheet",rows=50,cols=60)

Wall time: 422 ms


#### Pandas integration

In [31]:
#set the values of a pandas dataframe to sheet
wks.set_dataframe(df,start=(1,1))